src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,508 @@
+        {
+            if (bases.Length != qubits.Length)
+            {
+                throw new InvalidOperationException($"Both input arrays for Measure (bases, qubits), must be of same size.");
+            }
+
+            var qubitsToMeasure = bases.Zip(qubits, PauliFilter("Measure")).WhereNotNull();
+            var result = qubitsToMeasure.Aggregate(false, (accu, qubit) => accu ^ GetValue(qubit));
+            return result.ToResult();
+        }
+
+        // Overriding the Assert methods enables the use of statements such as
+        // `AssertQubit(Zero, q)` inside a Q# program.
+        public override void Assert(IQArray<Pauli> bases, IQArray<Qubit> qubits, Result expected, string msg)
+        {
+            if (bases.Length != qubits.Length)
+            {
+                throw new InvalidOperationException($"Both input arrays for Assert (bases, qubits), must be of same size.");
+            }
+
+            // All qubits are considered whose corresponding measurement basis
+            // is PauliZ.  Measurement in PauliI basis are ignored, and other
+            // bases will raise an exception.
+            var qubitsToMeasure = bases.Zip(qubits, PauliFilter("Assert")).WhereNotNull();
+
+            // A multi-qubit measurement in the PauliZ basis corresponds to
+            // computing the parity of all involved qubits' measurement values.
+            // (see also https://docs.microsoft.com/quantum/concepts/pauli-measurements#multiple-qubit-measurements)
+            // We use Aggregate to successively XOR a qubit's simulation value
+            // to an accumulator value `accu` that is initialized to `false`.
+            var actual = qubitsToMeasure.Aggregate(false, (accu, qubit) => accu ^ GetValue(qubit));
+            if (actual.ToResult() != expected)
+            {
+                // If the expected value does not correspond to the actual measurement
+                // value, we throw a Q# specific Exception together with the user
+                // defined message `msg`.
+                throw new ExecutionFailException(msg);
+            }
+        }
+
+        public override void AssertProb(IQArray<Pauli> bases, IQArray<Qubit> qubits, double probabilityOfZero, string msg, double tolerance)
+        {
+            if (bases.Length != qubits.Length)
+            {
+                throw new InvalidOperationException($"Both input arrays for AssertProb (bases, qubits), must be of same size.");
+            }
+
+            if ((probabilityOfZero + tolerance >= 1.0) && (probabilityOfZero - tolerance <= 0.0))
+            {
+                return;
+            }
+
+            // If there are any X- or Y-basis measurements, then the probability is 50%
+            if (bases.Any(p => p == Pauli.PauliX || p == Pauli.PauliY))
+            {
+                if (Abs(0.5 - probabilityOfZero) > tolerance)
+                {
+                    throw new ExecutionFailException(msg);
+                }
+            }
```

Comment:
I am confused. The processor only supports measurements in Z basis, right? So what does it mean to assert probability for X and Y?

Reply:
It's also not completely clear to me, but carried over from the conversion.